### ISSUE TITLE
[GPU Optimization] Hardcoded block=64 limits NVIDIA SM warp occupancy to ~50% (Ampere/Ada/Hopper)

### ISSUE THREAD

**Issue by bakiabaci** [enarjord/passivbot#1738]
### Component & Target
- **File:** `src/optimization/gpu/cuda_kernel.py`
- **Line:** 139 (`CudaShaderLibrary.__getattr__.launch`)
- **Category:** Performance Optimization / Hardware Throughput

---

### 1. Summary
The default CUDA thread block size in `cuda_kernel.py` is hardcoded to `block = 64`. While 64 is a standard threadgroup size for Apple Silicon Metal (where the shaders originated), on modern NVIDIA architectures (Ampere, Ada Lovelace, Hopper), a block size of 64 restricts hardware warp occupancy to **50%–75%**, leaving up to half of each Streaming Multiprocessor's (SM) execution units idle. 

Bumping the default to `block = 256` (or `128`) allows the GPU scheduler to reach **100% theoretical warp occupancy** without causing register spilling.

---

### 2. Architectural Root Cause & Occupancy Math

In NVIDIA GPUs, each Streaming Multiprocessor (SM) has hardware limits on both **maximum active threads (2048)** and **maximum active concurrent blocks**:

| Architecture | Max Active Threads / SM | Max Concurrent Blocks / SM | Warp Occupancy at `block = 64` | Warp Occupancy at `block = 256` |
| :--- | :--- | :--- | :--- | :--- |
| **Ampere (RTX 30xx, A100)** | 2048 | **16** | **50.0%** ($16 \times 64 = 1024$) | **100.0%** ($8 \times 256 = 2048$) |
| **Ada Lovelace (RTX 40xx)** | 2048 | **24** | **75.0%** ($24 \times 64 = 1536$) | **100.0%** ($8 \times 256 = 2048$) |
| **Hopper (H100)** | 2048 | **32** | **100.0%** ($32 \times 64 = 2048$) | **100.0%** ($8 \times 256 = 2048$) |

#### The Bottleneck on Ampere (e.g. RTX 3070/3080/A100):
Because the hardware caps concurrent blocks at **16 per SM**, running with `block = 64` means an SM can never host more than $16 \times 64 = 1024$ active threads. The remaining 1024 thread slots (and corresponding warp schedulers / FP32 ALUs) sit completely idle because the block scheduler cannot allocate more than 16 blocks.

#### Register File Overhead Check:
With `block = 256` (8 warps per block), reaching 2048 active threads utilizes 8 active blocks per SM. Assuming ~32 registers per thread:
$$2048 \text{ threads} \times 32 \text{ registers} = 65,536 \text{ registers per SM}$$
This matches NVIDIA's 64K 32-bit register file per SM exactly, meaning **zero register spilling to local DRAM**.

---

### 3. Surgical Diff Patch

```diff
--- a/src/optimization/gpu/cuda_kernel.py
+++ b/src/optimization/gpu/cuda_kernel.py
@@ -136,7 +136,10 @@ class CudaShaderLibrary:
             if not 1 <= count <= np.iinfo(np.uint32).max:
                 raise ValueError("CUDA dispatch must contain 1..2**32-1 threads")
             if group_size is not None and tuple(group_size[1:]) != (1, 1):
                 raise ValueError("CUDA proxy only supports one-dimensional thread groups")
-            block = 64 if group_size is None else int(group_size[0])
+            # Modern NVIDIA architectures (Ampere, Ada, Hopper) achieve optimal (100%)
+            # warp occupancy with 256 threads per block, preventing the hardware
+            # max-block limit per SM (e.g. 16 on Ampere) from artificially capping active threads.
+            block = 256 if group_size is None else int(group_size[0])
             if not 1 <= block <= 1024:
                 raise ValueError("Invalid CUDA thread block size")
```

---

### 4. Verification Plan

Verify candidate throughput improvements with Passivbot's proxy benchmark suite:

```bash
# Baseline vs optimized candidate evaluations/second:
python src/tools/gpu_proxy_benchmark.py --case all --candidates 1024 --warm-runs 10
python src/tools/gpu_proxy_benchmark.py --case tm-multicoin-overhead --candidates 2048 --coins 16 --warm-runs 10

# Ensure regression tests pass:
pytest pytests/ -k "gpu or cuda"
```

**Comment by bakiabaci** [comment]
In addition to the `block = 256` occupancy fix, we can also optimize the host dispatch loop in `src/optimization/gpu/cuda_kernel.py`.

Currently, on every kernel dispatch inside `CudaShaderLibrary.launch`:
1. Dynamic `torch.tensor([arg])` allocations are triggered for scalar parameters, causing repeated heap/VRAM allocation overhead.
2. A new `self._cupy.cuda.ExternalStream` wrapper object is instantiated on every launch.

Caching these scalar buffers and stream wrappers in `__init__` achieves a zero-allocation dispatch loop for hot optimization sweeps.

### Complete Patch:
```diff
--- a/src/optimization/gpu/cuda_kernel.py
+++ b/src/optimization/gpu/cuda_kernel.py
@@ -126,6 +126,8 @@ class CudaShaderLibrary:
 
         self._cupy = cupy
         self._scalar_parameters = {}
+        self._scalar_buffer_cache = {}
+        self._stream_cache = {}
         for name, signature in re.findall(r"kernel\s+void\s+(\w+)\((.*?)\)\s*\{", source, re.S):
             self._scalar_parameters[name] = {
                 index: match[1]
@@ -151,7 +153,7 @@ class CudaShaderLibrary:
                 raise ValueError("CUDA dispatch must contain 1..2**32-1 threads")
             if group_size is not None and tuple(group_size[1:]) != (1, 1):
                 raise ValueError("CUDA proxy only supports one-dimensional thread groups")
-            block = 64 if group_size is None else int(group_size[0])
+            block = 256 if group_size is None else int(group_size[0])
             if not 1 <= block <= 1024:
                 raise ValueError("Invalid CUDA thread block size")
             scalar_parameters = self._scalar_parameters.get(name, {})
@@ -167,16 +169,18 @@ class CudaShaderLibrary:
                     "float": torch.float32,
                     "bool": torch.bool,
                 }
-                args = tuple(
-                    (
-                        torch.tensor(
-                            [arg], device=device, dtype=scalar_types[scalar_parameters[index]]
-                        )
-                        if index in scalar_parameters and isinstance(arg, (int, float))
-                        else arg
-                    )
-                    for index, arg in enumerate(args)
-                )
+                new_args = []
+                for index, arg in enumerate(args):
+                    if index in scalar_parameters and isinstance(arg, (int, float)):
+                        cache_key = (name, index, arg, device)
+                        if cache_key not in self._scalar_buffer_cache:
+                            self._scalar_buffer_cache[cache_key] = torch.tensor(
+                                [arg], device=device, dtype=scalar_types[scalar_parameters[index]]
+                            )
+                        new_args.append(self._scalar_buffer_cache[cache_key])
+                    else:
+                        new_args.append(arg)
+                args = tuple(new_args)
             if not args or any(
                 not isinstance(arg, torch.Tensor)
                 or arg.device.type != "cuda"
@@ -189,9 +193,12 @@ class CudaShaderLibrary:
                 )
             device = args[0].device
             with torch.cuda.device(device), self._cupy.cuda.Device(device.index):
-                stream = self._cupy.cuda.ExternalStream(
-                    torch.cuda.current_stream(device).cuda_stream
-                )
+                cuda_stream_ptr = torch.cuda.current_stream(device).cuda_stream
+                if cuda_stream_ptr not in self._stream_cache:
+                    self._stream_cache[cuda_stream_ptr] = self._cupy.cuda.ExternalStream(
+                        cuda_stream_ptr
+                    )
+                stream = self._stream_cache[cuda_stream_ptr]
                 with stream:
                     arrays = tuple(self._cupy.from_dlpack(arg) for arg in args)
                     kernel(((count + block - 1) // block,), (block,), arrays + (np.uint32(count),))
```

All existing integration checks and test suites pass cleanly. Let me know if you would like a PR submitted for this!
